src/lib/staged_ledger/staged_ledger.ml

```diff
@@ -1215,7 +1215,118 @@ module T = struct
       =
     (List.map ~f:(With_status.map ~f:Transaction.forget) a, b, c, d)
 
-  let check_commands ledger ~verifier (cs : User_command.t With_status.t list) =
+  type transaction_pool_proxy =
+    { find_by_hash :
+           Mina_transaction.Transaction_hash.t
+        -> Mina_transaction.Transaction_hash.User_command_with_valid_signature.t
+           option
+    }
+
+  let dummy_transaction_pool_proxy : transaction_pool_proxy =
+    { find_by_hash = const None }
+
+  (* NOTE: when we have access to txn pool, any txns in pool should be already
+     verified, hence we could utilize this fact to perform a faster version of
+     command verification *)
+  let verify_command_against_pool
+      (transaction_pool_proxy : transaction_pool_proxy)
+      (cmd_with_status : User_command.Verifiable.t With_status.t) =
+    let With_status.{ data = verifiable_cmd; _ } = cmd_with_status in
+    let cmd_hash =
+      verifiable_cmd |> User_command.of_verifiable
+      |> Transaction_hash.hash_command
+    in
+    match transaction_pool_proxy.find_by_hash cmd_hash with
+    | None ->
+        Ok `No_fast_forward
+    | Some With_hash.{ data = cmd_in_pool; _ } -> (
+        (* NOTE:
```

Comment:
Let's move out this code to `Common` module under a new function `verify_commands_from_mempool`